HW2/P5/driver.py

```diff
@@ -0,0 +1,112 @@
+import sys
+import os.path
+sys.path.append(os.path.join('..', 'util'))
+
+import set_compiler
+set_compiler.install()
+
+import pyximport
+pyximport.install()
+
+import numpy as np
+import functools
+from timer import Timer
+from animator import Animator
+from physics import update, preallocate_locks
+
+
+def randcolor():
+    return np.random.uniform(0.0, 0.89, (3,)) + 0.1
+
+
+# From: https://en.wikipedia.org/wiki/Z-order_curve
+def cmp_zorder(a, b, dim=2):
+    j = 0
+    k = 0
+    x = 0
+    for k in range(dim):
+        y = a[k] ^ b[k]
+        if less_msb(x, y):
+            j = k
+            x = y
+    return int(a[j] - b[j])
+
+
+# From: https://en.wikipedia.org/wiki/Z-order_curve
+def less_msb(x, y):
+    return x < y and x < (x ^ y)
+
+
+if __name__ == '__main__':
+    num_balls = 500  # 500
+    radius = 0.01  # 0.01
+    positions = np.random.uniform(0 + radius, 1 - radius,
+                                  (num_balls, 2)).astype(np.float32)
+
+    # make a hole in the center
+    while True:
+        distance_from_center = np.sqrt(((positions - 0.5) ** 2).sum(axis=1))
+        mask = (distance_from_center < 0.25)
+        num_close_to_center = mask.sum()
+        if num_close_to_center == 0:
+            # everything is out of the center
+            break
+        positions[mask, :] = np.random.uniform(0 + radius, 1 - radius,
+                                               (num_close_to_center, 2)).astype(np.float32)
+
+    velocities = np.random.uniform(-0.25, 0.25,
+                                   (num_balls, 2)).astype(np.float32)
+
+    # Initialize grid indices:
+    #
+    # Each square in the grid stores the index of the object in that square, or
+    # -1 if no object.  We don't worry about overlapping objects, and just
+    # store one of them.
+    grid_spacing = radius / np.sqrt(2.0)
+    grid_size = int((1.0 / grid_spacing) + 1)
+    grid = - np.ones((grid_size, grid_size), dtype=np.uint32)
+    grid[(positions[:, 0] / grid_spacing).astype(int),
+         (positions[:, 1] / grid_spacing).astype(int)] = np.arange(num_balls)
+
+    # A matplotlib-based animator object
+    animator = Animator(positions, radius * 2)
+
+    # simulation/animation time variablees
+    physics_step = 1.0 / 100  # estimate of real-time performance of simulation
+    anim_step = 1.0 / 30  # FPS
+    total_time = 0
+
+    frame_count = 0
+
+    # SUBPROBLEM 4: uncomment the code below.
+    # preallocate locks for objects
+    locks_ptr = preallocate_locks(num_balls)
+
+    while True:
+        with Timer() as t:
+            update(positions, velocities, grid,
+                   radius, grid_size, locks_ptr,
+                   physics_step)
+
+            # Key and sort positions list for indexing
+            keyed_list = [(i, (positions[i][0], positions[i][1])) for i in xrange(num_balls)]
+            sorted_list = sorted(keyed_list, cmp=cmp_zorder, key=lambda x: (int(x[1][0] / grid_spacing), int(x[1][1] / grid_spacing)))
```

Comment:
I quite agree with amauboussin. Assigning a value to each cell of the grids seems to be easier than defining a comparison for this problem.
amauboussin mentioned 'approximately the right answer' and I do not think it is a serious problem. It is because what we are trying to achieve here is to make the geometric neighborhood also neighborhood in memory. The methods are never correct or wrong, but only good or bad.